.github/workflows/check.yml

```diff
@@ -1,53 +1,149 @@
+          skip_after_successful_duplicate: true
+          paths_ignore: '["**/README.md", "**/CHANGELOG.md", "**/LICENSE.md"]'
+          do_not_skip: '["pull_request"]'
+
   mix_check:
     name: mix check (Elixir ${{matrix.elixir}}, OTP ${{matrix.otp}})
+    needs: check_duplicate_runs
+    if: ${{ needs.check_duplicate_runs.outputs.should_skip != 'true' }}
 
     strategy:
       matrix:
-        include:
-          - elixir: '1.11.4'
-            otp: '23.2.7'
-          - elixir: '1.7.4'
-            otp: '20.3.8'
+        elixir:
+        - "1.7"
+        - "1.8"
+        - "1.9"
+        - "1.10"
+        - "1.11"
+        otp:
+        - "20.3"
+        - "21.0"
+        - "22.0.2"
+        - "23.0"
+        exclude:
+        - elixir: "1.10"
+          otp: "20.3"
+        - elixir: "1.11"
+          otp: "20.3"
 
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-18.04
 
     steps:
-      - name: Checkout
-        uses: actions/checkout@v2
+    - name: Checkout
+      uses: actions/checkout@v2
 
-      - name: Setup Elixir
-        uses: erlef/setup-elixir@v1
-        with:
-          elixir-version: ${{ matrix.elixir }}
-          otp-version: ${{ matrix.otp }}
+    - name: Set up Elixir
+      uses: erlef/setup-elixir@v1
+      with:
+        elixir-version: ${{ matrix.elixir }}
+        otp-version: ${{ matrix.otp }}
 
-      - name: Cache Elixir deps
-        uses: actions/cache@v2
-        with:
-          path: deps
-          key: elixir-deps-${{ runner.os }}-${{ hashFiles('mix.lock') }}
-          restore-keys: |
-            elixir-deps-${{ runner.os }}-
+    - name: Restore deps and _build cache
+      uses: actions/cache@v2
+      with:
+        path: |
+          deps
+          _build
+        key: deps-${{ runner.os }}-${{ matrix.otp }}-${{ matrix.elixir }}-${{ hashFiles('**/mix.lock') }}-git-${{ github.sha }}
+        restore-keys: |
+          deps-${{ runner.os }}-${{ matrix.otp }}-${{ matrix.elixir }}-${{ hashFiles('**/mix.lock') }}
+          deps-${{ runner.os }}-${{ matrix.otp }}-${{ matrix.elixir }}
 
-      - name: Fetch Elixir deps
-        run: mix deps.get
+    - name: Install package dependencies
+      run: mix deps.get
 
-      - name: Cache Elixir build
-        uses: actions/cache@v2
-        with:
-          path: _build
-          key: elixir-build-${{ runner.os }}-${{ matrix.otp }}-${{ matrix.elixir }}-${{ hashFiles('mix.lock') }}-${{ hashFiles('lib/**/*.ex', 'config/*.exs', 'mix.exs') }}
-          restore-keys: |
-            elixir-build-${{ runner.os }}-${{ matrix.otp }}-${{ matrix.elixir }}-${{ hashFiles('mix.lock') }}-
-            elixir-build-${{ runner.os }}-${{ matrix.otp }}-${{ matrix.elixir }}-
+    - name: Check Formatting
+      run: mix format --check-formatted
+
+    - name: Remove compiled application files
+      run: mix clean
+
+    - name: Compile dependencies
+      run: mix compile
+      env:
+        MIX_ENV: test
+
+    - name: Run check
+      run: mix check
+
+
+  dyalyzer:
```

Comment:
Why dialyzer? ex_check is a command line tool and doesn't expose any APIs that would benefits from typespecs, nor are they used internally. These runs will be completely redundant so it neglects the whole point of this PR - i.e. to optimize the CI workflows. Even if considering this file a reference for other ex_check users, I wouldn't want to teach them to blindly run all curated tools regardless of the project's needs (and .check.exs for this repo does disable dialyzer).

(+ typo in this line)

Reply:
This separated job was added because dialyzer check fails on old Elixir versions because some functions does not exist. But I found another solution for that, so this job is removed.